src/Npgsql/TypeMapping/ConnectorTypeMapper.cs

```diff
@@ -355,55 +367,115 @@ NpgsqlTypeHandler Bind(NpgsqlTypeMapping mapping, PostgresType? pgType = null)
         {
             pgType ??= GetPostgresType(mapping);
             var handler = mapping.TypeHandlerFactory.CreateNonGeneric(pgType, _connector);
-            Bind(handler, pgType, mapping.NpgsqlDbType, mapping.ClrTypes);
-
-            return handler;
+            return GetOrBind(handler, pgType, mapping.NpgsqlDbType, mapping.ClrTypes);
         }
 
-        void Bind(NpgsqlTypeHandler handler, PostgresType pgType, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null)
+        T GetOrBind<T>(T handler, PostgresType pgType, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null, bool isArray = false)
+            where T : NpgsqlTypeHandler
         {
-            if (_handlersByOID.TryGetValue(pgType.OID, out var existingHandler))
+            if (!TryBind(handler, pgType, out var existingHandler, npgsqlDbType, clrTypes, isArray))
             {
                 if (handler.GetType() != existingHandler.GetType())
                 {
                     throw new InvalidOperationException($"Two type handlers registered on same type OID '{pgType.OID}': " +
                                                         $"{existingHandler.GetType().Name} and {handler.GetType().Name}");
                 }
 
-                return;
+                // We can safely cast as we just compared a subtype of T (handler.GetType()) against existingHandler.GetType().
+                return (T)existingHandler;
             }
 
-            _handlersByOID[pgType.OID] = handler;
-            _handlersByTypeName[pgType.FullName] = handler;
-            _handlersByTypeName[pgType.Name] = handler;
+            return handler;
 
-            if (npgsqlDbType.HasValue)
+            bool TryBind(NpgsqlTypeHandler handler, PostgresType pgType, [NotNullWhen(false)]out NpgsqlTypeHandler? existingHandler, NpgsqlDbType? npgsqlDbType = null, Type[]? clrTypes = null, bool isArray = false)
             {
-                var value = npgsqlDbType.Value;
-                if (_handlersByNpgsqlDbType.ContainsKey(npgsqlDbType.Value))
+                var handlers = _handlers;
+                if (handlers.ByOID.TryGetValue(pgType.OID, out var lazy))
```

Comment:
Do we actually need this TryGetValue, considering that we do GetOrAdd just below? After all, due to a race condition we may get false here, but still just Get below - which you handle correctly, but which begs the question why we do this at all...

Reply:
Yeah I found it a bit questionable too but it avoids allocs from the GetorAdd, especially when we refactor this GetOrBind to be the top level call it's needed.

Reply:
Up to you... It's a bit confusing/redundant, so we can introduce it if/when we refactor GetOrBind to the top level, or if you prefer to leave it in also fine.